### PR TITLE
Therapie validation

### PR DESCRIPTION
**Pull request by dd52**
Rebased to develop and tested again using: 
1. https://cds.cds-test.staging.healex.systems/patient/mMtsl6rH/nNGM/export-td
2. https://cds.cds-test.staging.healex.systems/patient/mMtsl6rH/nNGM/export-td, with modified SD data as the export did not provide required uuids. 
3. Export data in test data/1. Therapie-CDS-Export-Original.json

### CODE REVIEW

**Review by ylboerner — COMMENTED** [review]
A few points regarding both compositions: 

1. Please set the code.coding to 1..1 for each therapy section
2. Could you set the cardinality for the 'entry' element in each section to either n (If all the entries are 1..1 anyway) or 1..n (we need the formularId in any case.) 
3. Remove the 'fallidentifier' from both compositions

I'll review the maps during the course of today.

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Operation.map [hunk 1]
```diff
@@ -15,7 +10,7 @@ uses "http://hl7.org/fhir/StructureDefinition/Patient" as target
 group TransformBundleOperationTherapie(source operations: BackboneElement, target bundle: Bundle, target composition: Composition, target section: BackboneElement)
 {
     operations then CreateProcedureOperationTherapie(operations, bundle, composition, section);
-    operations then CreatePatientOperationTherapie(operations, bundle, section);
+    // operations then CreatePatientOperationTherapie(operations, bundle, section);
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Operation.map:13]: Everything regarding the patient can be deleted

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map [hunk 2]
```diff
@@ -11,8 +11,8 @@ uses "http://hl7.org/fhir/StructureDefinition/Patient" as target
 group TransformBundleSonstigeTherapie(source operations: BackboneElement, target bundle: Bundle, target composition: Composition, target section: BackboneElement)
 {
     operations then CreateProcedureSonstigeTherapien(operations, bundle, composition, section);
-    operations then CreatePatientSonstigeTherapie(operations, bundle, composition, section);
-}
+    // operations then CreatePatientSonstigeTherapie(operations, bundle, composition, section);
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map:14]: Everything related to the patient can be deleted

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map [hunk 3]
```diff
@@ -203,7 +202,7 @@ group CreatePatientSonstigeTherapie(source operations: BackboneElement, target b
 group TransformPatientSonstigeTherapie(source operations: BackboneElement, target tgt: Patient, target composition: Composition, target section: BackboneElement)
 {
     // Mapped profile with profile URL
-    operations -> tgt.meta as meta collate, meta.profile = 'http://uk-koeln.de/fhir/StructureDefinition/Patient/nNGM';
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map:206]: The group for mapping the patient here can be deleted

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Strahlentherapie.map [hunk 4]
```diff
@@ -19,7 +15,7 @@ group TransformBundleStrahlentherapie(source operations: BackboneElement, target
 {
     operations then CreateProcedureStrahlentherapie(operations, bundle, composition, section);
     operations then CreateEpisodeOfCareStrahlentherapie(operations, bundle, composition, section);
-    operations then CreatePatientStrahlentherapie(operations, bundle, section);
+    // operations then CreatePatientStrahlentherapie(operations, bundle, section);
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Strahlentherapie.map:18]: Patient can be deleted here as well 👍

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map [hunk 5]
```diff
@@ -27,7 +14,7 @@ uses "http://hl7.org/fhir/StructureDefinition/Observation" as target
 group TransformBundleSystemischeTherapie(source operations: BackboneElement, target bundle: Bundle, target composition: Composition, target section: BackboneElement)
 {
     operations then CreateMedicationAdministrationSY(operations, bundle, composition, section);
-    operations then CreatePatientSY(operations, bundle, composition, section);
+    // operations then CreatePatientSY(operations, bundle, composition, section);
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map:17]: Can be deleted as well 👍

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map [hunk 6]
```diff
@@ -1,20 +1,7 @@
...
-        profile requires dosage.rate.value, .system and .code
-    - Missing valueSet for id_2342
-    - When 'Abbruch wegen Progress' option is checked in Art des Therapieendes (id_1356), it displays a field named 'Datum Progess' (blockindex = 6, groupindex = 0, itemid = 'sy_date_progress'),
-        but its information is not on Seatable nor Master excel file, so it's not being mapped
-    - id_1465 isn't mapping properly because it cannot override value already set by id_1246, even though it's effectivePeriod.start and id_1246 it's effective.start
-
-    EpisodeOfCare
-    - what to do with uuid_nngm_dateOfProgress (and/or uuid_nngm_dateOfProgress2)?
-*/
+// Todo: Extension(Date) needed for Datum Aenderung (id_1465)
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map:4]: Is this TODO still relevant? If so, let's solve it first 👍

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map [hunk 7]
```diff
@@ -538,7 +497,7 @@ group CreatePatientSY(source operations: BackboneElement, target bundle: Bundle,
 group TransformPatientSY(source operations: BackboneElement, target tgt: Patient, target composition: Composition, target section: BackboneElement)
 {
     // Profile url
-    operations -> tgt.meta as meta collate, meta.profile = 'http://uk-koeln.de/fhir/StructureDefinition/Patient/nNGM';
+    operations -> tgt.meta as meta collate, meta.profile = 'http://uk-koeln.de/fhir/StructureDefinition/Patient/nNGM/patient';
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map:500]: The transformation group for the patient can be deleted as well

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map [hunk 8]
```diff
@@ -1,282 +1,306 @@
 /// version = 0.1
 /// title = "nNGM: Operation Therapie - FHIR to CTS"
 
+// Art der Resektion, Resektionstatus not mapped in to FHIR Map due to collate bug
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map:4]: Is this still relevant? Let's solve this first 👍

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map [hunk 9]
```diff
@@ -1,282 +1,306 @@
...
+                procedure -> data.groupindex = 0;
+                procedure -> data.itemid = 'id_1162';
+                display -> data.values as values, values.value = display;
+            };  
+        };
+        
+        // Operative Intention
+        procedure -> tgt.data as data then
+        {
+            procedure.reasonCode as reasonCode, reasonCode.coding as coding where "coding.system = 'http://uk-koeln.de/fhir/CodeSystem/nngm/intention-therapie'" then
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map:44]: The blocks in this function could be simplified. The value is always the code except for 'N/A'. So basically, if it is != N/A, you could just map the value.

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Strahlentherapie.map [hunk 10]
```diff
@@ -1,431 +1,282 @@
 /// version = 0.1
 /// title = "nNGM: Sonstige Therapie - FHIR to CTS"
 
+// TODO: find out why output is produced twice
```
Comment (ylboerner) [inline CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Strahlentherapie.map:4]: Is this still relevant? If so, we should solve it first 👍

**Review by ylboerner — CHANGES_REQUESTED** [review]
Good job, the PR is already looking great. The comments directly in the code and the following issues are the only things left: 

**Extensions should only contain either an extension or a value, but never both**
If I transform this patient: https://cds.cds-test.staging.healex.systems/patient/8OH6ytqA/

I get the following error: 

image

This basically says, that you should not put extensions directly on the value of extensions, but rather on the extension itself. So an extension could never have a value and an extension at the same time.

In this specific case, the following profile http://uk-koeln.de/fhir/StructureDefinition/Procedure/nNGM/MedicationAdministration contains an extension, which has an extension on its valueBoolean. However, no valueBoolean should be present if the value is 'N/A' but only an extension on the extension should be set. 

**Entry cardinalities in the composition section**
The cardinalities for the entries in the sections are currently all set to 1..*. This is better than before, but they should actually represent the cardinalities of the entries. For example, if you have one formularid entry and one operation entry, the cardinality should be 2..2. Same for the other entries. 

**EpisodeOfCare missing in section for systemische Therapie**
The entry in the section for the EpisodeOfCare in the systemische Therapie is missing. 

**Pseudonymized and non pseudonymized compositions differ too much**
There are some min values and must support flags different in both compositions. The only cardinality difference should be in the non pseudonymized version, where the Stammdaten section is mandatory, so 1..1.

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Strahlentherapie.map (hunk 10 above)
Comment (dd52) [reply CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Strahlentherapie.map:4]: not relevant anymore

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map (hunk 9 above)
Comment (dd52) [reply CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map:41]: fixed!

File: CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map (hunk 8 above)
Comment (dd52) [reply CDS Maps/Individual Maps/FHIR-to-CDS/nNGM-FHIR-to-CDS-Operation.map]: not relevant anymore

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map (hunk 7 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map]: 410

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map (hunk 6 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map]: not relevant anymore

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map (hunk 5 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Systemische Therapie.map]: 410

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Strahlentherapie.map (hunk 4 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Strahlentherapie.map]: 410

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map (hunk 3 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map:206]: 410

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map (hunk 2 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Sonstige Therapie.map]: 410

File: CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Operation.map (hunk 1 above)
Comment (dd52) [reply CDS Maps/Individual Maps/CDS-to-FHIR/nNGM-CDS-to-FHIR-Operation.map]: 410

**Review by ylboerner — CHANGES_REQUESTED** [review]
No more validation errors, great job! There are only a few import issues left, but apart from that, it's looking awesome!

Below, you'll see a source patient on the left and the imported patient on the right. I check, whether they differ in any way. Here are some differences I found:

**1. Systemische Therapie: Therapie innerhalb einer klinischen Studie:**
Source: https://cds.cds-test.staging.healex.systems/patient/8OH6ytqA/nNGM/SY1
Target: https://cds.cds-test.staging.healex.systems/patient/CwkugMLQ/nNGM/SY1

image

The import is not working correctly yet. Maybe a twist with the boolean values?

**2. Systemische Therapie: Erhaltungstherapie**
Source: https://cds.cds-test.staging.healex.systems/patient/8OH6ytqA/nNGM/SY1
Target: https://cds.cds-test.staging.healex.systems/patient/CwkugMLQ/nNGM/SY1

image

Erhaltungstherapie is not mapping correctly.

**3. Systemische Therapie: Therapieschema**
Source: https://cds.cds-test.staging.healex.systems/patient/8OH6ytqA/nNGM/SY2
Target: https://cds.cds-test.staging.healex.systems/patient/CwkugMLQ/nNGM/SY2

image

**4. Sonstige Therapie: Therapiemerkmale**
Source: https://cds.cds-test.staging.healex.systems/patient/8OH6ytqA/nNGM/SO2
Target: https://cds.cds-test.staging.healex.systems/patient/CwkugMLQ/nNGM/SO2

image

The 'other' button is not working yet, although the text is mapped. So I guess it's just a matter of setting the value in the radio button.

**Pull request merged**